File: src/main/java/com/grosner/dbflow/sql/language/From.java
package com.grosner.dbflow.sql.language;

import com.grosner.dbflow.config.FlowManager;
import com.grosner.dbflow.sql.Query;
import com.grosner.dbflow.sql.builder.Condition;
import com.grosner.dbflow.sql.builder.ConditionQueryBuilder;
import com.grosner.dbflow.sql.builder.QueryBuilder;
import com.grosner.dbflow.structure.Model;

import java.util.ArrayList;
import java.util.List;

/**
 * Author: andrewgrosner
 * Contributors: { }
 * Description: The SQL FROM query wrapper that must have a {@link com.grosner.dbflow.sql.Query} base.
 */
public class From<ModelClass extends Model> implements WhereBase<ModelClass> {

    /**
     * The base such as {@link Delete}, {@link Select} and more!
     */
    private Query mQueryBuilderBase;

    /**
     * The table that this statement gets from
     */
    private Class<ModelClass> mTable;

    /**
     * An alias for the table
     */
    private String mAlias;

    /**
     * Enables the SQL JOIN statement
     */
    private List<Join> mJoins = new ArrayList<Join>();

    /**
     * The SQL from statement constructed.
     *
     * @param querybase The base query we append this query to
     * @param table     The table this corresponds to
     */
    public From(Query querybase, Class<ModelClass> table) {
        mQueryBuilderBase = querybase;
        mTable = table;
    }

    /**
     * The alias that this table name we use
     *
     * @param alias
     * @return
     */
    public From<ModelClass> as(String alias) {
        mAlias = alias;
        return this;
    }

    /**
     * Adds a join on a specific table for this query
     *
     * @param table    The table this corresponds to
     * @param joinType The type of join to use
     * @return
     */
    public <JoinType extends Model> Join<JoinType, ModelClass> join(Class<JoinType> table, Join.JoinType joinType) {
        Join<JoinType, ModelClass> join = new Join<JoinType, ModelClass>(this, table, joinType);
        mJoins.add(join);
        return join;
    }

    /**
     * Returns an empty {@link Where} statement
     *
     * @return
     */
    public Where<ModelClass> where() {
        return new Where<ModelClass>(this);
    }

    /**
     * Returns a {@link Where} statement with the sql clause
     *
     * @param whereClause The full SQL string after the WHERE keyword
     * @return
     */
    public Where<ModelClass> where(String whereClause) {
        return where().whereClause(whereClause);
    }

    /**
     * Returns a {@link Where} statement with the specified {@link com.grosner.dbflow.sql.builder.ConditionQueryBuilder}
     *
     * @param conditionQueryBuilder
     * @return
     */
    public Where<ModelClass> where(ConditionQueryBuilder<ModelClass> conditionQueryBuilder) {
        return where().whereQuery(conditionQueryBuilder);
    }

    /**
     * Returns a {@link Where} statement with the specified array of {@link com.grosner.dbflow.sql.builder.Condition}
     *
     * @param conditions The array of conditions that define this WHERE statement
     * @return
     */
    public Where<ModelClass> where(Condition... conditions) {
        return where().andThese(conditions);
    }

    public Set<ModelClass> set() {
        if (!(mQueryBuilderBase instanceof Update)) {
            throw new IllegalStateException("Cannot use set() without an UPDATE as the base");
        }
        return new Set<ModelClass>(this, mTable);
    }

    public Set<ModelClass> set(Condition... conditions) {
        return set().conditions(conditions);
    }

    public Set<ModelClass> set(ConditionQueryBuilder<ModelClass> conditionQueryBuilder) {
        return set().conditionQuery(conditionQueryBuilder);
    }

    @Override
    public String getQuery() {
        QueryBuilder queryBuilder = new QueryBuilder()
                .append(mQueryBuilderBase.getQuery());
        if(!(mQueryBuilderBase instanceof Update)) {
            queryBuilder.append("FROM ");
        }

        queryBuilder.append(FlowManager.getTableName(mTable));

        if (mQueryBuilderBase instanceof Select) {
            queryBuilder.appendSpace().appendQualifier("AS", mAlias);
            for (Join join : mJoins) {
                queryBuilder.append(join.getQuery());
            }
        } else {
            queryBuilder.appendSpace();
        }

        return queryBuilder.getQuery();
    }

    @Override
    public String toString() {
        return getQuery();
    }

    /**
     * The base query
     *
     * @return
     */
    public Query getQueryBuilderBase() {
        return mQueryBuilderBase;
    }

    /**
     * The table this From corresponds to
     *
     * @return
     */
    public Class<ModelClass> getTable() {
        return mTable;
    }
}

File: lib/src/main/kotlin/com/dbflow5/query/From.kt
package com.dbflow5.query

import com.dbflow5.adapter.SQLObjectAdapter
import com.dbflow5.config.FlowManager
import com.dbflow5.query.Join.JoinType
import com.dbflow5.query.property.IndexProperty
import com.dbflow5.sql.Query
import com.dbflow5.structure.ChangeAction
import kotlin.reflect.KClass
import kotlin.collections.Set as KSet

/**
 * Description: The SQL FROM query wrapper that must have a [Query] base.
 */
class From<TModel : Any>
/**
 * The SQL from statement constructed.
 *
 * @param queryBuilderBase The base query we append this cursor to
 * @param table     The table this corresponds to
 */
internal constructor(

    /**
     * @return The base query, usually a [Delete], [Select], or [Update]
     */
    override val queryBuilderBase: Query,
    adapter: SQLObjectAdapter<TModel>,

    /**
     * If specified, we use this as the subquery for the FROM statement.
     */
    private val modelQueriable: ModelQueriable<TModel>? = null
) : BaseTransformable<TModel>(adapter) {

    /**
     * An alias for the table
     */
    private var tableAlias: NameAlias? = null

    /**
     * Enables the SQL JOIN statement
     */
    private val joins = mutableListOf<Join<*, *>>()

    override val primaryAction: ChangeAction
        get() = if (queryBuilderBase is Delete) ChangeAction.DELETE else ChangeAction.CHANGE

    override val query: String
        get() {
            val queryBuilder = StringBuilder()
                .append(queryBuilderBase.query)
            if (queryBuilderBase !is Update<*>) {
                queryBuilder.append("FROM ")
            }

            modelQueriable?.let { queryBuilder.append(it.enclosedQuery) }
                ?: queryBuilder.append(getTableAlias())

            if (queryBuilderBase is Select) {
                if (!joins.isEmpty()) {
                    queryBuilder.append(" ")
                }
                joins.forEach { queryBuilder.append(it.query) }
            } else {
                queryBuilder.append(" ")
            }

            return queryBuilder.toString()
        }

    override fun cloneSelf(): From<TModel> {
        val from = From(
            when (queryBuilderBase) {
                is Select -> queryBuilderBase.cloneSelf()
                else -> queryBuilderBase
            },
            adapter as SQLObjectAdapter<TModel>,
        )
        from.joins.addAll(joins)
        from.tableAlias = tableAlias
        return from
    }

    /**
     * @return A list of [Class] that represents tables represented in this query. For every
     * [Join] on another table, this adds another [Class].
     */
    val associatedTables: KSet<KClass<*>>
        get() = linkedSetOf<KClass<out Any>>(table).apply {
            joins.mapTo(this) { it.adapter.table }
        }

    private fun getTableAlias(): NameAlias = tableAlias
        ?: NameAlias.Builder(FlowManager.getTableName(table)).build().also { tableAlias = it }

    /**
     * Set an alias to the table name of this [From].
     */
    infix fun `as`(alias: String): From<TModel> {
        tableAlias = getTableAlias()
            .newBuilder()
            .`as`(alias)
            .build()
        return this
    }

    /**
     * Adds a join on a specific table for this query
     *
     * @param table    The table this corresponds to
     * @param joinType The type of join to use
     */
    fun <TJoin : Any> join(
        adapter: SQLObjectAdapter<TJoin>,
        joinType: JoinType
    ): Join<TJoin, TModel> {
        return Join(this, adapter, joinType).also { joins.add(it) }
    }

    /**
     * Adds a join on a specific table for this query.
     *
     * @param modelQueriable A query we construct the [Join] from.
     * @param joinType       The type of join to use.
     */
    fun <TJoin : Any> join(
        modelQueriable: ModelQueriable<TJoin>,
        joinType: JoinType
    ): Join<TJoin, TModel> {
        return Join(this, joinType, modelQueriable).also { joins.add(it) }
    }

    /**
     * Adds a [JoinType.CROSS] join on a specific table for this query.
     *
     * @param table   The table to join on.
     * @param <TJoin> The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> crossJoin(adapter: SQLObjectAdapter<TJoin>): Join<TJoin, TModel> =
        join(adapter, JoinType.CROSS)

    /**
     * Adds a [JoinType.CROSS] join on a specific table for this query.
     *
     * @param modelQueriable The query to join on.
     * @param <TJoin>        The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> crossJoin(modelQueriable: ModelQueriable<TJoin>): Join<TJoin, TModel> =
        join(modelQueriable, JoinType.CROSS)

    /**
     * Adds a [JoinType.INNER] join on a specific table for this query.
     *
     * @param table   The table to join on.
     * @param <TJoin> The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> innerJoin(adapter: SQLObjectAdapter<TJoin>): Join<TJoin, TModel> =
        join(adapter, JoinType.INNER)

    /**
     * Adds a [JoinType.INNER] join on a specific table for this query.
     *
     * @param modelQueriable The query to join on.
     * @param <TJoin>        The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> innerJoin(modelQueriable: ModelQueriable<TJoin>): Join<TJoin, TModel> =
        join(modelQueriable, JoinType.INNER)

    /**
     * Adds a [JoinType.LEFT_OUTER] join on a specific table for this query.
     *
     * @param table   The table to join on.
     * @param <TJoin> The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> leftOuterJoin(adapter: SQLObjectAdapter<TJoin>): Join<TJoin, TModel> =
        join(adapter, JoinType.LEFT_OUTER)

    /**
     * Adds a [JoinType.LEFT_OUTER] join on a specific table for this query.
     *
     * @param modelQueriable The query to join on.
     * @param <TJoin>        The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> leftOuterJoin(modelQueriable: ModelQueriable<TJoin>): Join<TJoin, TModel> =
        join(modelQueriable, JoinType.LEFT_OUTER)


    /**
     * Adds a [JoinType.NATURAL] join on a specific table for this query.
     *
     * @param adapter   The table to join on.
     * @param <TJoin> The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> naturalJoin(adapter: SQLObjectAdapter<TJoin>): Join<TJoin, TModel> =
        join(adapter, JoinType.NATURAL)

    /**
     * Adds a [JoinType.NATURAL] join on a specific table for this query.
     *
     * @param modelQueriable The query to join on.
     * @param <TJoin>        The class of the join table.
    </TJoin> */
    infix fun <TJoin : Any> naturalJoin(modelQueriable: ModelQueriable<TJoin>): Join<TJoin, TModel> =
        join(modelQueriable, JoinType.NATURAL)

    /**
     * Begins an INDEXED BY piece of this query with the specified name.
     *
     * @param indexProperty The index property generated.
     */
    infix fun indexedBy(indexProperty: IndexProperty<TModel>): IndexedBy<TModel> =
        IndexedBy(indexProperty, this)

}

/**
 * Extracts the [From] from a [ModelQueriable] if possible to get [From.associatedTables]
 */
@Suppress("UNCHECKED_CAST")
fun <T : Any> ModelQueriable<T>.extractFrom(): From<T>? {
    return if (this is From<*>) {
        this as From<T>
    } else if (this is Where<*> && whereBase is From<*>) {
        whereBase as From<T>
    } else {
        null
    }
}
